src/Raven.Server/Documents/Queries/Graph/GraphQueryRunner.cs

```diff
@@ -0,0 +1,193 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+using Raven.Client.Extensions;
+using Raven.Server.Documents.Queries.AST;
+using Raven.Server.Documents.Queries.Dynamic;
+using Raven.Server.ServerWide;
+using Raven.Server.ServerWide.Context;
+using Sparrow;
+using Sparrow.Json;
+
+namespace Raven.Server.Documents.Queries.Graph
+{
+    public class GraphQueryRunner
+    {
+        protected readonly DocumentDatabase Database;
+        protected readonly GraphQuery Query;
+
+        public GraphQueryRunner([NotNull] DocumentDatabase database, [NotNull] GraphQuery q)
+        {
+            Database = database ?? throw new ArgumentNullException(nameof(database));
+            Query = q ?? throw new ArgumentNullException(nameof(q));
+        }
+
+        public async Task<DocumentQueryResult> RunAsync()
+        {
+            var withQueryRunner = new DynamicQueryRunner(Database);
+            
+            //first do the with clause queries
+            using (Database.DocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext ctx))
+            using (ctx.OpenReadTransaction())
+            {
+                var withClauseResults = new Dictionary<string, List<Match>>();
+                foreach (var withClause in Query.WithDocumentQueries)
+                {
+                    var withClauseQueryResult =
+                        await withQueryRunner.ExecuteQuery(new IndexQueryServerSide(withClause.Value.ToString()), ctx, null, OperationCancelToken.None);
+                    var queryMatches = new List<Match>();
+                    foreach (var result in withClauseQueryResult.Results)
+                    {
+                        var match = new Match();
+                        match.Set(withClause.Key, result.Data);
```

Comment:
I'm trying to follow the logic here, you add the key of the clause which I assume is the alias and the document data but the Match.Set will add this to a dictionary and this is done per Query result so i think Match.Set will throw if you add two documents to the same alias (I didn't run the code)

Reply:
Also, I wonder about adding those blittables to the Dictionary while the context is been dispose of very soon after, shouldn't we either clone the data or have a longer living context here?